pkg/controllers/allocation/scheduling/constraints.go

```diff
@@ -15,56 +15,138 @@ limitations under the License.
+func generateTaints(constraints *v1alpha4.Constraints, pod *v1.Pod) error {
+	taints := scheduling.Taints(constraints.Taints)
 	for _, toleration := range pod.Spec.Tolerations {
-		// Only OpEqual is supported
+		// Only OpEqual is supported. OpExists does not make sense for
+		// provisioning -- in theory we could create a taint on the node with a
+		// random string, but it's unclear use case this would accomplish.
 		if toleration.Operator != v1.TolerationOpEqual {
 			continue
 		}
+		var generated []v1.Taint
 		// Use effect if defined, otherwise taint all effects
 		if toleration.Effect != "" {
-			overrides = append(overrides, v1.Taint{Key: toleration.Key, Value: toleration.Value, Effect: toleration.Effect})
+			generated = []v1.Taint{{Key: toleration.Key, Value: toleration.Value, Effect: toleration.Effect}}
 		} else {
-			overrides = append(overrides,
-				v1.Taint{Key: toleration.Key, Value: toleration.Value, Effect: v1.TaintEffectNoSchedule},
-				v1.Taint{Key: toleration.Key, Value: toleration.Value, Effect: v1.TaintEffectNoExecute},
-			)
+			generated = []v1.Taint{
+				{Key: toleration.Key, Value: toleration.Value, Effect: v1.TaintEffectNoSchedule},
+				{Key: toleration.Key, Value: toleration.Value, Effect: v1.TaintEffectNoExecute},
+			}
 		}
+		// Only add taints that do not already exist on constraints
+		for _, taint := range generated {
+			if !taints.Has(taint) {
+				taints = append(taints, taint)
+			}
+		}
+	}
+	constraints.Taints = taints
+	return nil
+}
+
+func generateLabels(constraints *v1alpha4.Constraints, pod *v1.Pod) error {
+	labels := map[string]string{}
+	// Default to constraint labels
+	for key, value := range constraints.Labels {
+		labels[key] = value
+	}
+	// Override with pod labels
+	nodeAffinity := scheduling.NodeAffinityFor(pod)
+	for _, key := range nodeAffinity.GetLabels() {
+		if _, ok := v1alpha4.WellKnownLabels[key]; !ok {
+			var labelConstraints []string
+			if value, ok := constraints.Labels[key]; ok {
+				labelConstraints = append(labelConstraints, value)
+			}
+			values := nodeAffinity.GetLabelValues(key, labelConstraints)
+			if len(values) == 0 {
+				return fmt.Errorf("label %s is too constrained", key)
+			}
+			labels[key] = values[0]
+		}
+	}
+	constraints.Labels = labels
+	return nil
+}
+
+func validateTopology(pod *v1.Pod) (errs error) {
+	for _, constraint := range pod.Spec.TopologySpreadConstraints {
+		if supported := []string{v1.LabelHostname, v1.LabelTopologyZone}; !functional.ContainsString(supported, constraint.TopologyKey) {
+			errs = multierr.Append(errs, fmt.Errorf("unsupported topology key, %s not in %s", constraint.TopologyKey, supported))
+		}
+	}
+	return errs
+}
+
+func validateAffinity(pod *v1.Pod) (errs error) {
+	if pod.Spec.Affinity == nil {
+		return nil
+	}
+	if pod.Spec.Affinity.PodAffinity != nil {
+		errs = multierr.Append(errs, fmt.Errorf("pod affinity is not supported"))
+	}
+	if pod.Spec.Affinity.PodAntiAffinity != nil {
```

Comment:
Until we respect the constraints, we basically *have* to not schedule the pods. I can imagine eventually trying to implement pod affinity, but minimally it will be incremental effort over this. FWIW, the current behavior is to not provision if constraints are not supported.